src/loaders/compute_tools/microtrait/microtrait_util.py

```diff
@@ -20,9 +23,53 @@
     _UNWRAPPED_RULE_COL]
 
 
+def _retrieve_trait_substrate_mapping(
+        substrate2rule_file: Path
+) -> dict[str, set[str]]:
+    # retrieve the trait to substrate mapping from the substrate2rule file
+    substrate2rule_df = pd.read_csv(substrate2rule_file, sep='\t')
+
+    # create a dictionary of trait names to substrate names from the substrate2rule file
+    trait_substrate_mapping = defaultdict(set)
+    for index, row in substrate2rule_df.iterrows():
+        substrate_name = row[_SUBSTRATE2RULE_SUBSTRATE_NAME_COL]
+        for trait_name_col in _RULE2TRAIT_TRAIT_NAME_COLS:
+            trait_name = row[trait_name_col]
+            trait_substrate_mapping[trait_name].add(substrate_name) if pd.notna(trait_name) else None
+
+    return trait_substrate_mapping
+
+
+def _retrieve_genes(
+        unwrapped_rule: str
+) -> set[str]:
+    # retrieve and extract gene names from the unwrapped rule expression.
+    # i.e. "('pcbA' | 'pcbB' | 'pcbC' | 'pcbD' | 'pcbE' | 'pcbF' | 'pcbG' | 'pcbH')", "((((((('tc.fev')))))))",
+    # "(('sreA' & 'sreB' & 'sreC') | ('hydG' & 'hydB_4'))", etc.
+
+    # find gene names between single quotes
+    gene_names = set(re.findall(r"'([^']+)'", unwrapped_rule))
+
+    return gene_names
+
+
+def _write_trait_unwrapped_rules(
```

Comment:
These aren't really rules at this point, they're just gene / hmm names

Reply:
👍

Reply:
:+1: